OMCompiler/SimulationRuntime/c/simulation/solver/gbode_util.c

```diff
@@ -0,0 +1,544 @@
+    }
+  }
+}
+
+/**
+ * @brief         Linear interpolation of specific vector components
+ *
+ * @param ta      Time value at the left hand side
+ * @param fa      Function values at the left hand side
+ * @param tb      Time value at the right hand side
+ * @param fb      Function values at the right hand side
+ * @param t       Time value at the interpolated time point
+ * @param f       Function values at the interpolated time point
+ * @param nIdx    Size of index vector
+ * @param idx     Index vector
+ */
+void linear_interpolation_gbf(double ta, double* fa, double tb, double* fb, double t, double* f, int nIdx, int* idx)
+{
+  double lambda, h0, h1;
+  int i, ii;
+
+  if (tb == ta) {
+    // omit division by zero
+    copyVector_gbf(f, fb, nIdx, idx);
+  } else {
+    lambda = (t-ta)/(tb-ta);
+    h0 = 1-lambda;
+    h1 = lambda;
+
+    for (ii=0; ii<nIdx; ii++)
+    {
+      i = idx[ii];
+      f[i] = h0*fa[i] + h1*fb[i];
+    }
+  }
+}
+
+/**
+ * @brief Hermite interpolation of specific vector components
+ *
+ * @param ta      Time value at the left hand side
+ * @param fa      Function values at the left hand side
+ * @param dfa     Derivative function values at the left hand side
+ * @param tb      Time value at the right hand side
+ * @param fb      Function values at the right hand side
+ * @param dfb     Derivative function values at the right hand side
+ * @param t       Time value at the interpolated time point
+ * @param f       Function values at the interpolated time point
+ * @param nIdx    Size of index vector
+ * @param idx     Index vector
+ */
+void hermite_interpolation_gbf(double ta, double* fa, double* dfa, double tb, double* fb, double* dfb, double t, double* f, int nIdx, int* idx)
+{
+  double tt, h00, h01, h10, h11;
+  int i, ii;
+
+  if (tb == ta) {
+    // omit division by zero
+    copyVector_gbf(f, fb, nIdx, idx);
+  } else {
+    tt = (t-ta)/(tb-ta);
+    h00 = (1+2*tt)*(1-tt)*(1-tt);
+    h10 = (tb-ta)*tt*(1-tt)*(1-tt);
+    h01 = (3-2*tt)*tt*tt;
+    h11 = (tb-ta)*(tt-1)*tt*tt;
+
+    for (ii=0; ii<nIdx; ii++)
+    {
+      i = idx[ii];
+      f[i] = h00*fa[i]+h10*dfa[i]+h01*fb[i]+h11*dfb[i];
+    }
+  }
+}
+
+/**
+ * @brief Difference between linear and hermite interpolation at intermediate points
+ *
+ * @param ta      Time value at the left hand side
+ * @param fa      Function values at the left hand side
+ * @param dfa     Derivative function values at the left hand side
+ * @param tb      Time value at the right hand side
+ * @param fb      Function values at the right hand side
+ * @param dfb     Derivative function values at the right hand side
+ * @param t       Time value at the interpolated time point
+ * @param f       Function values at the interpolated time point
+ * @param nIdx    Size of index vector
+ * @param idx     Index vector
+ */
+void error_interpolation_gbf(double ta, double* fa, double* dfa, double tb, double* fb, double* dfb, double t, double* f, int nIdx, int* idx)
```

Comment:
It may be faster to compute the error directly, but t would be safer to call both interpolation methods and then compute only the error.

Otherwise each change has to be made here as well.